analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs

```diff
@@ -433,6 +435,76 @@ public void Branching_LearnsObjectConstraint_TypePattern_Negated(string expressi
                 .And.ContainSingle(x => x != null && x.HasConstraint(ObjectConstraint.NotNull));
         }
 
+        [DataTestMethod]
+        [DataRow("arg is { }")]
+        [DataRow("arg is { }", "TClass")]
+        [DataRow("arg is object { }")]
+        [DataRow("arg is not not { }")]
+        [DataRow("!!(arg is { })")]
+        [DataRow("arg is (A: var a, B: _)", "Deconstructable")]
+        public void Branching_LearnsObjectConstraint_RecursivePattern_ElseIsNull(string expression, string argType = "object")
+        {
+            var validator = CreateIfElseEndValidatorCS(expression, OperationKind.RecursivePattern, argType);
+            validator.ValidateTag("If", x => x.HasConstraint(ObjectConstraint.NotNull).Should().BeTrue());
+            validator.ValidateTag("Else", x => x.HasConstraint(ObjectConstraint.Null).Should().BeTrue());
+            validator.TagValues("End").Should().HaveCount(2)
+                .And.ContainSingle(x => x.HasConstraint(ObjectConstraint.Null))
+                .And.ContainSingle(x => x.HasConstraint(ObjectConstraint.NotNull));
+        }
+
+        [DataTestMethod]
+        [DataRow("arg is string { }")]
+        [DataRow("arg is string { Length: 0 }")]
+        [DataRow("arg is string { Length: var length }")]
+        public void Branching_LearnsObjectConstraint_RecursivePattern_ElseIsUnknown(string expression)
+        {
+            var validator = CreateIfElseEndValidatorCS(expression, OperationKind.RecursivePattern);
+            validator.ValidateTag("If", x => x.HasConstraint(ObjectConstraint.NotNull).Should().BeTrue());
+            validator.ValidateTag("Else", x => x.Should().BeNull("it could be null or any other type"));
+            validator.TagValues("End").Should().HaveCount(2)
+                .And.ContainSingle(x => x == null)
+                .And.ContainSingle(x => x != null && x.HasConstraint(ObjectConstraint.NotNull));
+        }
+
+        [DataTestMethod]
+        [DataRow("arg is { }", "int")]
+        [DataRow("arg is { }", "T")]
```

Comment:
In the true branch, we know that T is not null. Nullable analysis thinks so too:
![image](https://user-images.githubusercontent.com/103252490/188118723-9fc000b2-31af-418e-a92f-b98d39166484.png)

https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4GIYFcA2OLA4CmABETAcQLABQAAgMwl0BMJAwiQN60l/NM6AFhIBZADwAVAHwAKSSQgAnAOYBKbr37aoAMxKzlKklADO3EgF81W7Xx407TxaoB0kgPYBlAC5LYKrJqANy2dpZhETSWQA
```suggestion
```

Reply:
T is not constraint and doesnt' have `IsReferenceType`. I won't fix it